prog/base.rb

```diff
@@ -179,11 +179,20 @@ def donate
   end
 
   def reap
-    exited_children = strand.children_dataset.where(Sequel.~(exitval: nil)).all
-
-    # Clear cache if anything was deleted.
-    strand.associations.delete(:children) if exited_children.count > 0
-    exited_children.each(&:destroy)
+    deleted = strand.children.filter_map { |child|
+      next unless child.exitval
+
+      # Clear any semaphores that get added to a exited Strand prog,
+      # since incr is entitled to be run at *any time* (including
+      # after exitval is set, though it doesn't do anything) and any
+      # such incements will prevent deletion of a Strand via
+      # foreign_key
+      Semaphore.where(strand_id: child.id).destroy
+      child.destroy
+    }
+
+    strand.children.delete_if { deleted.include?(_1) }
```

Comment:
Not sure about precisely the best way to do this, but I think this is alright?

Reply:
@enescakir maybe you know something about this? I think you've hacked Sequel extension the most (though, Burak found that one surprising behavior in it, too....)

This is downstream of our sometimes beneficial but a bit fidgety attempt to reduce the number of reloads while keeping the in-memory representation somewhat current.

Reply:
This should be the array delete_if, to remove elements. There's a destroy, for database stuff, above, but I'm trying to keep the association array in sync with changes I know have been applied